### PR TITLE
Add local first architecture to React SDK

### LINKED ISSUE #48: Feature request: Use DB first approach
### Is your feature request related to a problem?

This feature is related to performance of client applications. More info here: https://xmtp.org/docs/tutorials/performance

### Describe the solution to the problem

This feature requires syncing messages to a local database layer that is then used as the source of truth to serve content from the XMTP network.

### Describe the uses cases for the feature

For some of the current and upcoming content types, this feature will be required to provide a user experience that's on par with current messaging platforms.

### Additional details

_No response_

### PR DESCRIPTION
**Pull request by rygine**
This is the first in a series of PRs related to the upcoming production release of the React SDK.

The biggest changes of this PR intend to resolve #48, which brings a local-first architecture to the React SDK. All of these changes can be found in the [Add local first architecture to React SDK](https://github.com/xmtp/xmtp-web/pull/50/commits/8c0fcc784893d277df076a92a181bb1fada384f7) commit.

### Chores

This PR also contains several maintenance commits to upgrade dependencies and optimize the DX.

### Updated components and example app

Both the components package and the example app now support most of our official-track content types. Still some work to be done here, but nothing super important at this time.

### CODE REVIEW

File: packages/react-sdk/src/contexts/XMTPContext.tsx [hunk 1]
```diff
@@ -62,6 +129,93 @@ export const XMTPProvider: React.FC<React.PropsWithChildren> = ({
   const [error, setError] = useState<unknown | null>(null);
   const initializingRef = useRef(false);
 
+  // combine all processors into a single object
+  const processors = useMemo(() => {
+    const finalCacheConfig = [...defaultCacheConfig, ...(cacheConfig ?? [])];
+    return {
```
Comment (daria-github) [inline packages/react-sdk/src/contexts/XMTPContext.tsx:135]: do you want to add some unit tests for this logic? a bit hard to read, esp with the nested reduce.

File: packages/react-sdk/src/helpers/isValidAddress.ts [hunk 2]
```diff
@@ -1,5 +1,8 @@
 /**
  * Check if a wallet address is valid.
+ *
+ * @param address The string to validate
```
Comment (daria-github) [inline packages/react-sdk/src/helpers/isValidAddress.ts:4]: super nit for `address The`

File: packages/react-sdk/src/hooks/useConversation.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+/**
+ * This hook returns helper functions for working with conversations in the
+ * local cache.
+ */
+export const useConversation = () => {
+  const { client } = useClient();
+  const { db } = useDb();
+
+  /* c8 ignore start */
```
Comment (daria-github) [inline packages/react-sdk/src/hooks/useConversation.ts:25]: nit: might be worth just adding a comment explaining why you're c8 ignoring

File: packages/react-sdk/src/helpers/caching/contentTypes/text.test.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
+import { it, expect, describe, vi } from "vitest";
+import { Client, ContentTypeText } from "@xmtp/xmtp-js";
+// eslint-disable-next-line import/no-extraneous-dependencies
```
[the remark this answers is not in the document]
Comment (rygine) [reply packages/react-sdk/src/helpers/caching/contentTypes/text.test.ts:3]: yep, i should add test files to the exceptions for this rule

File: packages/react-sdk/src/contexts/XMTPContext.tsx (hunk 1 above)
Comment (rygine) [reply packages/react-sdk/src/contexts/XMTPContext.tsx:135]: agreed! i basically ran out of time for testing and had to pick and choose my battles for a release. i can do this in a fast-follow.

**Review by daria-github — DISMISSED** [review]

**Review by daria-github — APPROVED** [review]

**Pull request merged**